test/CMakeLists.txt

```diff
@@ -0,0 +1,12 @@
+if (NOT GC_TEST_ENABLE)
+    message(STATUS "The tests are not enabled.")
+    return()
+endif ()
+
+include(gtest)
+
+# Static library to be used in tests
+add_library(graph_compiler_static STATIC ${GC_LIB_SOURCES})
```

Comment:
A question regarding gtest: is it compatible with LLVM's lit+gtest-unittest? We may need some C++-based unittests for MLIR in the future. They may use `add_mlir_unittest` in cmake which is provided in MLIR upstream, to register a unittest in the llvm-lit system. `add_mlir_unittest` will make the target depend on LLVM's copy of gtest.

Reply:
definition of "add_mlir_unittest" seems trivial:
```cmake
function(add_mlir_unittest test_dirname)
  add_unittest(MLIRUnitTests ${test_dirname} ${ARGN})
endfunction()
```
And as far as I know lit tests and untitests (aka gtest) are not connected. Both of them used in llvm/mlir and unittests located in some folder like: `llvm-project/mlir/unittests/Analysis/Presburger`, but lit is `llvm-project/mlir/test/Transforms/control-flow-sink.mlir` and they have different cmakes and so on. 
What do you mean by lit+gtest?

Reply:
> definition of "add_mlir_unittest" seems trivial:

Please look deeper into the definition of `add_unittest`: it is not provided by cmake. Instead, it is defined in AddLLVM.cmake of upstream LLVM. It sets some path like the config path for `lit`.

> And as far as I know lit tests and untitests (aka gtest) are not connected

No, it's not true. Please try the cmake target `check-mlir` or  `check-mlir-unit` in upstream LLVM/MLIR. It will build and run the unittest (in the context of MLIR, they call gtest "unittest"). I think `llvm-lit` is the top-level test driver for all of `FileCheck` `Python` and `gtests`.
I did some research on how unittest is linked with lit. Besides creating executables via `add_unittest`, MLIR upstream adds the unittest cmake-build target to `MLIR_TEST_DEPENDS` at `mlir/test/CMakeLists.txt`. The targets in `MLIR_TEST_DEPENDS` will be marked as dependency at `add_lit_testsuite` below.

Also, MLIR upstream has a directory at `mlir/test/Unit`. The `lit.cfg.py` in it indicates that it is a GoogleTest suite. "add_mlir_unittest" seems to internally adjust the output directory to make sure the cfg is generated at the right place.

Reply:
Okay, now I get your point. I've added lit tests in the similar way with tpp repo - https://github.com/intel/graph-compiler/pull/33. 

I guess for the first period of time we will just use lit and run gtests separately. I think we should understand how many mlir/llvm unittests we need and if there are will be many of them we should try to use this approach. 

I agree with you, but we should back to this when we will have understanding of which tests we need.